src/coreclr/tools/Common/Internal/Runtime/EETypeBuilderHelpers.cs

```diff
@@ -136,6 +142,23 @@ private static bool HasCriticalFinalizer(TypeDesc type)
             return false;
         }
 
+        private static bool IsTrackedReferenceWithFinalizer(TypeDesc type)
+        {
+            do
+            {
+                if (!type.HasFinalizer)
+                    return false;
+
+                if (type is MetadataType mdType && mdType.HasCustomAttribute("System.Runtime.InteropServices.ObjectiveC", "ObjectiveCTrackedTypeAttribute"))
```

Comment:
Nit: would be odd to have a non-MetadataType (i.e. array/pointer/fnptr/...) with a finalizer.

```suggestion
                if ((MetadataType)type).HasCustomAttribute("System.Runtime.InteropServices.ObjectiveC", "ObjectiveCTrackedTypeAttribute"))
```